README.md

```diff
@@ -331,6 +331,51 @@ export default Modifier.extend({
 
 Whenever possible, it is recommended that you use the default "modern" API instead of the classic API.
 
+## TypeScript
+
+Using the "modern" native class API, you can use `.ts` instead of `.js` and it'll just work, as long as you do runtime checks to narrow the types of your args when you access them. 
+
+```ts
+// app/modifiers/scroll-position.ts
+import Modifier from 'ember-class-based-modifier';
+
+export default class ScrollPositionModifier extends Modifier {
+  // ...
+}
+```
+
+But to avoid writing runtime checks, you can extend `Modifier` with your own args, similar to the way you would define your args for a Glimmer Component. 
+
+```ts
+// app/modifiers/scroll-position.ts
+import Modifier from 'ember-class-based-modifier';
+
+interface ScrollPositionModifierArgs {
+  positional: number[],
+  // this modifier doesn't accept named args, but this is included as an example:
+  named: {
+    delay: number
+  }
+}
+
+export default class ScrollPositionModifier extends Modifier<ScrollPositionModifierArgs> {
+  get scrollPosition(): number {
+    // get the first positional argument passed to the modifier
+    //
+    // {{scoll-position @someNumber}}
+    //                  ~~~~~~~~~~~
+    //
+    return this.args.positional[0];
```

Comment:
@chriskrycho is that your general philosophy/recommendation that this is "fine"?

Specifically, since we don't have "typed handlebars", the general arg is really just *asserting* to TypeScript that "I somehow know this is a number and I don't want a type error", when you really don't (the caller could have passed you a number, a string, or forgot to pass it). In Ember/Glimmer we always stick to `unknown[]` and `Dict<unknown>` and do debug-mode runtime checks if we care, or explicit any cast if we really want to say that.

That being said, Glimmer components also have the generic arg and therefore also the same "problem", so maybe this is the pragmatic choice the Ember-TS community has come to terms with?

Reply:
Yes, you're quite correct about the state of things at the moment. I'll quote myself from the original PR here:

> Note here that TS users will likely want to use debug `assert` on those arg types in the constructor until we have a solution in place for type-checking template invocation, which is unlikely to happen this calendar year.

This approach strikes a nice balance between the ergonomics throughout the body of a component or helper—you don't have to validate the args in every single computed property definition, for example—and validating them to make sure the type you wrote is actually what you got handed. So, in the example code used in the README:

```diff
export default class ScrollPositionModifier extends Modifier<ScrollPositionModifierArgs> {
  get scrollPosition(): number {
    return this.args.positional[0];
  }

  didReceiveArguments() {
+   assert(typeof this.args.positional[0] === 'number');
+   assert(typeof this.args.named.delay === 'number');

    this.element.scrollTop = this.scrollPosition;
  }
}
```

Reply:
Note as well: @pzuraq and I and a few others have been batting around some template-friendly and class-friendly approaches to `argTypes`, which would solve this quite nicely in its own way both for general JS users *and* as an interim for TS users until we're able to build type-checking-for-templates. Hopefully over the next ~year, we'll see incremental but very real improvements in the experience of authoring with types in this space!

Reply:
I thought about adding this to the example in the readme, but I'm not sure if it is helpful without the context you gave about templates. So instead I linked to your comment on my old PR. If you think it makes sense in the readme, I'm happy to add it.

Reply:
I think leaving it with that link for now is fine. If we see further confusion comes up, we can address it then, but this is more just something the entire Ember TS community has to know to deal with, not specific to this addon.

Reply:
Isn’t named getters (where you can put the runtime assertions or type assertions/casts) a better pattern to avoid the repetition?

Reply:
Having the generics doesn’t prevent you from doing that so I’m fine with it, but I’m now wondering whether there are any advantages to recommending that at all or just tell people to use the getter pattern all the time.

Reply:
There are roughly the following ways to solve this, at least that come to mind (if you have more, I'd love to hear them!):

1. **Validate args or do runtime type narrowing in every named getter which references a given argument.** You can certainly make this *less* verbose by extracting the validation to functions… but then you need to make sure that your build pipeline has the right sequencing to make sure they can be tree-shaken (which ours *should* as tree-shaking lands, but it's definitely a thing we'll have to keep an eye out for).

2. **Provide named getters for each argument, which themselves do the validation; never reference `this.args.foo` or `@foo`.** This certainly works, and it is better than (1), but it means that you now need to use *those* instead of the named arguments themselves, both in the JS and in the template. This loses you a lot of the clarity of named args in both JS and template contexts, too! Worse, this increases bundle size and memory allocation costs. That's not a *large* cost, but spread over an app the size of some of the larger ones out there, it really does matter quite a bit.

3. **Put all the assertions on arguments in one place, an appropriate lifecycle hook/modifier.** This has several advantages:
    - Assuming good decisions around that guidance of what lifecycle to put it on, it means that validation happens *when the arg changes*, and not lazily when the template happens to request the value referenced by the getter. This is apt to bite you in more complex components, where some of the properties might not be rendered and therefore might not be pulled by the template… so you have a bad argument, but it gets missed until a much later point in the lifecycle, possibly somewhere further down a tree of components (maybe not even *your* components!).

        (Thinking about this increases my sense of urgency around `argTypes` and particularly the ability to do that in dev-only contexts in templates. Especially as we move toward a templates-first, and much more often templates-only world, we'll need to build some more dev-only infra there.)

    - If/when the time comes that we have type-checked templates and/or a good `argTypes` implementation, there's exactly and only one place to (re)move things – that one lifecycle context. It's also much clearer when *reading* a component exactly what you're seeing: the place where it receives updates to arguments, you validate they are what you expect.

    - Done in the class with `assert`, there's *zero* production runtime overhead to them—not carrying around extra named getters on the class just for this purpose. If we can build some dev-only tools in templates that are analogous, that'll be similarly a great win for large apps.

Like I said, that's off the top of my head, so it's possible I missed something. This is just what I've landed on after working with a number of different approaches; I'm sure it's by no means exhaustive!

Reply:
@chriskrycho I think I'm basically proposing 2, at least for the named arguments you are using in TypeScript land, and I think that also counts as "put all the assertions on arguments in one place".

Glimmer components doesn't really have lifecycle hooks (or at least they were in the process of being deprecated/removed), so I think you have other choices there. Consuming things overly eagerly is also somewhat an anti-pattern imo, as it messes up auto-tracking etc.

I don't think this is really about providing a full-on runtime type checking solution; this is really just narrowly about bridging untyped code (the caller) with TypeScript code (the class body). Template-land is already untyped and I think we can just come to terms with it and work on template type checking another day, so using `@foo` in the template "unsafely" doesn't really bother me, personally. (As opposed to using `this.foo` in templates, which I would not want.) Either the next layer will catch it if they care, or it would be no worse than what you have to do to debug it without TS.

I don't think the code size angle really matters. Whether you stripe the runtime check or not, you still only have exactly one copy of the check. Sure.. you have a few more bytes of `get foo() { ... }` (~10 bytes or so per name) to ship... but on the flip side, `this.args.foo` becomes `this.foo` and neither can be minified, so if you reference it a few times you will come out ahead 🤷‍♂. I'm just not too worried about it.

Aside: I am not particularly fond of the `argTypes` helper solution, at least that's my first initial thought. It changes the semantics of `@foo` in noticeable ways which I think will cause problems.

Reply:
Anyway, I agree that this is something we have already chosen to do on a wider scale across the ecosystem so it's not really related to this PR or this addon anymore. We can find a different venue for that discussion somewhere else.